docs/configuration.md
`config/initializers/react_on_rails_pro.rb`

1. You don't need to create a initializer if you are satisfied with the defaults as described below.
1. Values beginning with `renderer` pertain only to using an external rendering server. You will need to ensure these values are consistent with your configuration for the external rendering server, as given in [docs/node-renderer/js-configuration.md](./node-renderer/js-configuration.md)
1. `config.prerender_caching` works for standard mini_racer server rendering and using an external rendering server.

# Example of Configuration

Also see [spec/dummy/config/initializers/react_on_rails_pro.rb](../../spec/dummy/config/initializers/react_on_rails_pro.rb) for how the testing app is setup.

The below example is a typical production setup, using the separate `NodeRenderer`, where development takes the defaults when the ENV values are not specified.

```ruby
ReactOnRailsPro.configure do |config|
  # If true, then capture timing of React on Rails Pro calls including server rendering and
  # component rendering.
  # Default for `tracing` is false.
  config.tracing = true

  # Array of globs to find any files for which changes should bust the fragment cache for
  # cached_react_component and cached_react_component_hash. This should include any files used to
  # generate the JSON props, webpack and/or webpacker configuration files, and npm package lockfiles.
  # Default for `dependency_globs` is an empty array
  config.dependency_globs = [ File.join(Rails.root, "app", "views", "**", "*.jbuilder") ]

  # Array of globs to exclude from config.dependency_globs for ReactOnRailsPro cache key hashing
  # Default for `excluded_dependency_globs` is an empty array
  config.excluded_dependency_globs = [ File.join(Rails.root, "app", "views", "**", "dont_hash_this.jbuilder") ]

  # Remote bundle caching saves deployment time by caching bundles.
  # See /docs/bundle-caching.md for usage and an example of a module called S3BundleCacheAdapter.
  config.remote_bundle_cache_adapter = nil
  
  # ALL OPTIONS BELOW ONLY APPLY IF SERVER RENDERING

  # If true, then cache the evaluation of JS for prerendering using the standard Rails cache.
  # Applies to all rendering engines.
  # Default for `prerender_caching` is false.  
  config.prerender_caching = true

  # Retry request in case of time out on the node-renderer side
  # 5 - default, if not specified
  # 0 - no retry
  config.renderer_request_retry_limit = 5

  # NodeRenderer is for a renderer that is stateless. It does not need restarting when the JS bundles
  # are updated. It is the only custom renderer currently supported. Leave blank to use the standard
  # mini_racer rendering. Other option is NodeRenderer
  # Default for `server_renderer` is "ExecJS"
  config.server_renderer = "NodeRenderer"

  # React on Rails Node Renderer now support render functions returning promises! To enable this optional functionality,
  # toggle the following option.
  # Default is false.
  config.rendering_returns_promises = false

  # If you're using the NodeRenderer, a value of true allows errors to be thrown from the bundle
  # code for SSR so that an error tracking system on the NodeRender can use the exceptions.
  # If you are using ExecJS as your rendering method, set this to false.
  # Default is true.
  config.throw_js_errors = true

  # You may provide a password and/or a port that will be sent to renderer for simple authentication.
  # `https://:<password>@url:<port>`. For example: https://:myPassword1@renderer:3800. Don't forget
  # the leading `:` before the password. Your password must also not contain certain characters that
  # would break calling URI(config.renderer_url). This includes: `@`, `#`, '/'.
  # **Note:** Don't forget to set up **SSL** connection (https) otherwise password will useless
  # since it will be easy to intercept it.
  # If you provide an ENV value (maybe only for production) and there is no value, then you get the default.
  # Default for `renderer_url` is "http://localhost:3800".
  config.renderer_url = ENV["RENDERER_URL"]

  # If you don't want to worry about special characters in your password within the url, use this config value
  # Default for `renderer_password` is ""
  # config.renderer_password = ENV["RENDERER_PASSWORD"]

  # Set the `ssr_timeout` configuration so the Rails server will not wait more than this many seconds
  # for a SSR request to return once issued. 
  config.ssr_timeout = 5
  
  # If false, then crash if no backup rendering when the remote renderer is not available
  # Can be useful to set to false in development or testing to make sure that the remote renderer
  # works and any non-availability of the remote renderer does not just do ExecJS.
  # Suggest setting this to false if the SSR JS code cannot run in ExecJS
  # Default for `renderer_use_fallback_exec_js` is false.
  config.renderer_use_fallback_exec_js = false

  # The maximum size of the http connection pool,
  # Set +pool_size+ to limit the maximum number of connections allowed.
  # Defaults to 1/4 the number of allowed file handles.  You can have no more
  # than this many threads with active HTTP transactions.
  # Default for `renderer_http_pool_size` is 10
  config.renderer_http_pool_size = 10

  # Seconds to wait for an available connection before a Timeout::Error is raised
  # Default for `renderer_http_pool_timeout` is 5
  config.renderer_http_pool_timeout = 5

  # warn_timeout  - Displays an error message if a checkout takes longer that the given time in seconds
  # (used to give hints to increase the pool size). Default is 0.25
  config.renderer_http_pool_warn_timeout = 0.25 # seconds

  # Snippet of JavaScript to be run right at the beginning of the server rendering process. The code
  # to be executed must either be self contained or reference some globally exposed module.  
  # For example, suppose that we had to call `SomeLibrary.clearCache()`between every call to server
  # renderer to ensure no leakage of state between calls. Note, SomeLibrary needs to be globally
  # exposed in the server rendering webpack bundle. This code is visible in the tracing of the calls
  # to do server rendering. Default is nil.
  config.ssr_pre_hook_js = "SomeLibrary.clearCache();"

  # When using the Node Renderer, you may require some extra assets in addition to the bundle.
  # The assets_to_copy option allows the Node Renderer to have assets copied at the end of
  # the assets:precompile task or directly by the
  # react_on_rails_pro:copy_assets_to_remote_vm_renderer task.
  # These assets are also transferred any time a new bundle is sent from Rails to the renderer.
  # The value should be a file_path or an Array of file_paths. The files should have extensions
  # to resolve the content types, such as "application/json".
  config.assets_to_copy = [
     Rails.root.join("public", "webpack", Rails.env, "loadable-stats.json"),
     Rails.root.join("public", "webpack", Rails.env, "manifest.json")
  ]
end
```


docs/vm-renderer/js-configuration.md
# VM Renderer JavaScript Configuration

You can configure the vm-renderer with only ENV values using the provided bin file `vm-renderer`.

You can also create a custom configuration file to setup and launch the vm-renderer.

The values in this file must be kept in sync with with the `config/initializers/react_on_rails_pro.rb` file, as documented in [docs/configuration.md](../configuration.md).

Here are the options available for the JavaScript renderer configuration object, as well as the available default ENV values if using the command line program vm-renderer.

1. **port** (default: `process.env.RENDERER_PORT || 3800`) - The port renderer should listen to. 
   If setting the port, you might want to ensure the port uses `process.env.PORT` so it will use port number provided by **Heroku** environment. 
1. **logLevel** (default: `process.env.RENDERER_LOG_LEVEL || 'info'`) - Log lever for renderer. Set it to `'error'` to turn logging off. Available levels are: `{ error: 0, warn: 1, info: 2, verbose: 3, debug: 4, silly: 5 }`
1. **bundlePath** (default: `process.env.RENDERER_BUNDLE_PATH || '/tmp/react-on-rails-pro-vm-renderer-bundles'` ) - path to temp directory where uploaded bundle files will be stored. For example you can set it to `path.resolve(__dirname, './tmp/bundles')` if you configured renderer from the `/` directory of your app. 
1. **workersCount** (default: `env.RENDERER_WORKERS_COUNT || defaultWorkersCount()` where default is your CPUs count - 1) - Number of workers that will be forked to serve rendering requests. If you set this manually make sure that value is a **Number** and is `>= 1`.
1. **password** (default: `env.RENDERER_PASSWORD`) - Password expected to receive form **Rails client** to authenticate rendering requests. If no password set, no authentication will be required.
1. **allWorkersRestartInterval** (default: `env.RENDERER_ALL_WORKERS_RESTART_INTERVAL`) - Interval in minutes between scheduled restarts of all cluster of workers. By default restarts are not enabled. If restarts are enabled, `delayBetweenIndividualWorkerRestarts` should also be set.
1. **delayBetweenIndividualWorkerRestarts** (default: `env.RENDERER_DELAY_BETWEEN_INDIVIDUAL_WORKER_RESTARTS`) - Interval in minutes between individual worker restarts (when cluster restart is triggered). By default restarts are not enabled. If restarts are enabled, `allWorkersRestartInterval` should also be set.
1. **honeybadgerApiKey** - (default: `env.HONEYBADGER_API_KEY`) - If you want errors on the VM Renderer to be sent to Honeybadger, set this value.
1. **supportModules** - (default: `env.RENDERER_SUPPORT_MODULES`) - Set this to true if you want the
   renderer to support Node modules, such as for `loadable-components`. Set to null if you want the execJS behavior.

## Example Launch Files

### Testing example: 

[spec/dummy/client/vm-renderer.js](../../spec/dummy/client/vm-renderer.js)

### Simple example:

Create a file './vm-renderer.js'
```js
import path from 'path';
import { reactOnRailsProVmRenderer } from 'react-on-rails-pro-vm-renderer';

const config = {
  // Save bundles to relative "./tmp/bundles" dir of our app 
  bundlePath: path.resolve(__dirname, './tmp/bundles'), 
  
  // All other values are the defaults, as described above 
};

// Renderer detects a total number of CPUs on virtual hostings like Heroku or CircleCI instead
// of CPUs number allocated for current container. This results in spawning many workers while
// only 1-2 of them really needed.
if (process.env.CI) {
  config.workersCount = 2;
}

reactOnRailsProVmRenderer(config);

```

And add this line to your `scripts` section of `package.json`

```json
  "scripts": {
    "start": "echo 'Starting React on Rails Pro VM Renderer.' && node ./vm-renderer.js"
  },
```

`yarn start` will run the renderer.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. Items under `Unreleased` is upcoming features that will be out in next release.

## Gem and Package Versions
Gem and package versions are the same except for beta releases where the gem uses a `.beta` and the package uses a `-beta` (same for `rc`).

1. **Gem**: `3.0.0.rc.1`
2. **Package**: `3.0.0-rc.1`

You can find the **package** version numbers from this repo's tags and below in this file.

----

[HEAD compared to 3.2.1]: Click to see all changes compared to the last 3.x version. See the [4.0 Release Notes](https://github.com/shakacode/react_on_rails_pro/blob/master/docs/release-notes/4.0.md).

## [Unreleased]
*Add changes in master not yet tagged.*

### Changed
- Set `bodyLimit` to 100 MB by default to fix error 413. [PR 511](https://github.com/shakacode/react_on_rails_pro/pull/511) by [Romex91](https://github.com/Romex91)

### Fixed
- Enabled `queueMicrotask` use in server bundle to support React 19. [PR 505](https://github.com/shakacode/react_on_rails_pro/pull/505) by [alexeyr-ci](https://github.com/alexeyr-ci).

### Changed
- Renamed `includeTimerPolyfills` configuration option to `stubTimers`. [PR 506](https://github.com/shakacode/react_on_rails_pro/pull/506) by [alexeyr-ci](https://github.com/alexeyr-ci).
- Fail immediately on detecting obsolete config options to prevent unexpected misconfigurations. [PR 506](https://github.com/shakacode/react_on_rails_pro/pull/506) by [alexeyr-ci](https://github.com/alexeyr-ci).

## [4.0.0.rc.10] - 2025-01-06

### Fixed
- Enable use as a `git:` dependency. [PR 490](https://github.com/shakacode/react_on_rails_pro/pull/490) by [alexeyr-ci](https://github.com/alexeyr-ci).

## [4.0.0.rc.9] - 2024-12-05

### Changed
- Error reporting and tracing integrations are completely redone. See [the docs](./docs/node-renderer/error-reporting-and-tracing.md) for details. [PR 471](https://github.com/shakacode/react_on_rails_pro/pull/471) by [alexeyr-ci](https://github.com/alexeyr-ci).
- Upgraded to Fastify 5 by default, with an option to fall back to Fastify 4 on older Node versions. [PR 478](https://github.com/shakacode/react_on_rails_pro/pull/478) by [alexeyr-ci](https://github.com/alexeyr-ci).
- Logging now uses Pino instead of Winston, aligning with Fastify. [PR 479](https://github.com/shakacode/react_on_rails_pro/pull/479) by [alexeyr-ci](https://github.com/alexeyr-ci).
  - In particular, this can make a difference on uncaught exceptions and unhandled rejections; now the process should always exit if they happen, as recommended.
- Converted from `Net::HTTP` to HTTPX for requests to the Node renderer [PR 452](https://github.com/shakacode/react_on_rails_pro/pull/452) by [alexeyr-ci](https://github.com/alexeyr-ci).
- Communication with Node Renderer now uses HTTP/2 Cleartext [PR 392](https://github.com/shakacode/react_on_rails_pro/pull/392) by [alexeyr-ci](https://github.com/alexeyr-ci).

## [4.0.0.rc.6] - 2024-11-12

### Added
- Added streaming server rendering support:
  - [PR 407](https://github.com/shakacode/react_on_rails_pro/pull/407) by [AbanoubGhadban](https://github.com/AbanoubGhadban).
    - New `stream_view_containing_react_components` helper method that can be used with `stream_react_component` helper method in react_on_rails gem.
    - Enables progressive page loading and improved performance for server-rendered React components.
  - Added support for replaying console logs that occur during server rendering of streamed React components. This enables debugging of server-side rendering issues by capturing and displaying console output on the client and on the server output. [PR #429](https://github.com/shakacode/react_on_rails_pro/pull/429) by [AbanoubGhadban](https://github.com/AbanoubGhadban).
  - Added support for handling errors happening during server rendering of streamed React components [PR #432](https://github.com/shakacode/react_on_rails_pro/pull/432) by [AbanoubGhadban](https://github.com/AbanoubGhadban):
    - It handles errors that happen during the initial render and errors that happen inside suspense boundaries.
    - Added `raise_non_shell_server_rendering_errors` configuration option.
- Added support for replaying console logs from asynchronous operations:
  - New `replayServerAsyncOperationLogs` configuration option to enable/disable this feature
  - When enabled, captures and replays console output from async operations during server-side rendering
  - [PR 440](https://github.com/shakacode/react_on_rails_pro/pull/440) by [AbanoubGhadban](https://github.com/AbanoubGhadban)

### Removed
- Drop support for EOL'd Ruby 2.7 [PR 365](https://github.com/shakacode/react_on_rails_pro/pull/365) by [ahangarha](https://github.com/ahangarha).
- Drop support for React on Rails below 14.0.4 [PR 415](https://github.com/shakacode/react_on_rails_pro/pull/415) by [rameziophobia](https://github.com/rameziophobia).

### Fixed
- Updated multiple JS dependencies for bug fixes.
- Added execute permission for `spec/dummy/bin/dev` [PR 387](https://github.com/shakacode/react_on_rails_pro/pull/387) by [alexeyr](https://github.com/alexeyr).
- Made default bundle paths in node-renderer and Rails consistent [PR 399](https://github.com/shakacode/react_on_rails_pro/pull/399) by [alexeyr-ci](https://github.com/alexeyr-ci).

### Changed
- Converted JS code to TS [PR 386](https://github.com/shakacode/react_on_rails_pro/pull/386) and [PR 389](https://github.com/shakacode/react_on_rails_pro/pull/389) by [alexeyr-ci](https://github.com/alexeyr-ci).
- Support Shakapacker 8.0.0, Modified webpack configurations to use shakapacker instead of webpacker. This drops support for shakapacker 6.X [PR 415](https://github.com/shakacode/react_on_rails_pro/pull/415) by [rameziophobia](https://github.com/rameziophobia).
- Converted the node-renderer worker from Express to Fastify [PR 398](https://github.com/shakacode/react_on_rails_pro/pull/398) by [alexeyr-ci](https://github.com/alexeyr-ci).


lib/react_on_rails_pro/configuration.rb
# frozen_string_literal: true

module ReactOnRailsPro
  def self.configure
    yield(configuration)
    configuration.setup_config_values
  end

  def self.configuration
    @configuration ||= Configuration.new(
      prerender_caching: Configuration::DEFAULT_PRERENDER_CACHING,
      server_renderer: Configuration::DEFAULT_RENDERER_METHOD,
      renderer_url: Configuration::DEFAULT_RENDERER_URL,
      renderer_use_fallback_exec_js: Configuration::DEFAULT_RENDERER_FALLBACK_EXEC_JS,
      renderer_http_pool_size: Configuration::DEFAULT_RENDERER_HTTP_POOL_SIZE,
      renderer_http_pool_timeout: Configuration::DEFAULT_RENDERER_HTTP_POOL_TIMEOUT,
      renderer_http_pool_warn_timeout: Configuration::DEFAULT_RENDERER_HTTP_POOL_WARN_TIMEOUT,
      renderer_password: nil,
      tracing: Configuration::DEFAULT_TRACING,
      dependency_globs: Configuration::DEFAULT_DEPENDENCY_GLOBS,
      excluded_dependency_globs: Configuration::DEFAULT_EXCLUDED_DEPENDENCY_GLOBS,
      remote_bundle_cache_adapter: Configuration::DEFAULT_REMOTE_BUNDLE_CACHE_ADAPTER,
      ssr_timeout: Configuration::DEFAULT_SSR_TIMEOUT,
      ssr_pre_hook_js: nil,
      assets_to_copy: nil,
      renderer_request_retry_limit: Configuration::DEFAULT_RENDERER_REQUEST_RETRY_LIMIT,
      throw_js_errors: Configuration::DEFAULT_THROW_JS_ERRORS,
      rendering_returns_promises: Configuration::DEFAULT_RENDERING_RETURNS_PROMISES,
      profile_server_rendering_js_code: Configuration::DEFAULT_PROFILE_SERVER_RENDERING_JS_CODE,
      raise_non_shell_server_rendering_errors: Configuration::DEFAULT_RAISE_NON_SHELL_SERVER_RENDERING_ERRORS
    )
  end

  class Configuration
    DEFAULT_RENDERER_URL = "http://localhost:3800"
    DEFAULT_RENDERER_METHOD = "ExecJS"
    DEFAULT_RENDERER_FALLBACK_EXEC_JS = true
    DEFAULT_RENDERER_HTTP_POOL_SIZE = 10
    DEFAULT_RENDERER_HTTP_POOL_TIMEOUT = 5
    DEFAULT_RENDERER_HTTP_POOL_WARN_TIMEOUT = 0.25
    DEFAULT_SSR_TIMEOUT = 5
    DEFAULT_PRERENDER_CACHING = false
    DEFAULT_TRACING = false
    DEFAULT_DEPENDENCY_GLOBS = [].freeze
    DEFAULT_EXCLUDED_DEPENDENCY_GLOBS = [].freeze
    DEFAULT_REMOTE_BUNDLE_CACHE_ADAPTER = nil
    DEFAULT_RENDERER_REQUEST_RETRY_LIMIT = 5
    DEFAULT_THROW_JS_ERRORS = true
    DEFAULT_RENDERING_RETURNS_PROMISES = false
    DEFAULT_PROFILE_SERVER_RENDERING_JS_CODE = false
    DEFAULT_RAISE_NON_SHELL_SERVER_RENDERING_ERRORS = false
    DEFAULT_ENABLE_RSC_SUPPORT = false
    DEFAULT_RSC_RENDERING_URL_PATH = "rsc/"

    attr_accessor :renderer_url, :renderer_password, :tracing,
                  :server_renderer, :renderer_use_fallback_exec_js, :prerender_caching,
                  :renderer_http_pool_size, :renderer_http_pool_timeout, :renderer_http_pool_warn_timeout,
                  :dependency_globs, :excluded_dependency_globs, :rendering_returns_promises,
                  :remote_bundle_cache_adapter, :ssr_pre_hook_js, :assets_to_copy,
                  :renderer_request_retry_limit, :throw_js_errors, :ssr_timeout,
                  :profile_server_rendering_js_code, :raise_non_shell_server_rendering_errors, :enable_rsc_support,
                  :rsc_rendering_url_path

    def initialize(renderer_url: nil, renderer_password: nil, server_renderer: nil, # rubocop:disable Metrics/AbcSize
                   renderer_use_fallback_exec_js: nil, prerender_caching: nil,
                   renderer_http_pool_size: nil, renderer_http_pool_timeout: nil,
                   renderer_http_pool_warn_timeout: nil, tracing: nil,
                   dependency_globs: nil, excluded_dependency_globs: nil, rendering_returns_promises: nil,
                   remote_bundle_cache_adapter: nil, ssr_pre_hook_js: nil, assets_to_copy: nil,
                   renderer_request_retry_limit: nil, throw_js_errors: nil, ssr_timeout: nil,
                   profile_server_rendering_js_code: nil, raise_non_shell_server_rendering_errors: nil,
                   enable_rsc_support: nil, rsc_rendering_url_path: nil)
      self.renderer_url = renderer_url
      self.renderer_password = renderer_password
      self.server_renderer = server_renderer
      self.renderer_use_fallback_exec_js = renderer_use_fallback_exec_js
      self.prerender_caching = prerender_caching
      self.renderer_http_pool_size = renderer_http_pool_size
      self.renderer_http_pool_timeout = renderer_http_pool_timeout
      self.renderer_http_pool_warn_timeout = renderer_http_pool_warn_timeout
      self.tracing = tracing
      self.rendering_returns_promises = server_renderer == "NodeRenderer" ? rendering_returns_promises : false
      self.dependency_globs = dependency_globs
      self.excluded_dependency_globs = excluded_dependency_globs
      self.remote_bundle_cache_adapter = remote_bundle_cache_adapter
      self.ssr_pre_hook_js = ssr_pre_hook_js
      self.assets_to_copy = assets_to_copy
      self.renderer_request_retry_limit = renderer_request_retry_limit
      self.throw_js_errors = throw_js_errors
      self.ssr_timeout = ssr_timeout
      self.profile_server_rendering_js_code = profile_server_rendering_js_code
      self.raise_non_shell_server_rendering_errors = raise_non_shell_server_rendering_errors
      self.enable_rsc_support = enable_rsc_support
      self.rsc_rendering_url_path = rsc_rendering_url_path
    end

    def setup_config_values
      configure_default_url_if_not_provided
      validate_url
      validate_remote_bundle_cache_adapter
      setup_renderer_password
      setup_assets_to_copy
      setup_execjs_profiler_if_needed
    end

    def setup_execjs_profiler_if_needed
      return unless profile_server_rendering_js_code && server_renderer == "ExecJS"

      if ExecJS.runtime == ExecJS::Runtimes::Node
        ExecJS.runtime = ExecJS::ExternalRuntime.new(
          name: "Node.js (V8)",
          command: ["node --prof"],
          runner_path: "#{ExecJS.root}/support/node_runner.js",
          encoding: "UTF-8"
        )
      elsif ExecJS.runtime == ExecJS::Runtimes::V8
        ExecJS.runtime = ExecJS::ExternalRuntime.new(
          name: "V8",
          command: ["d8 --prof"],
          runner_path: "#{ExecJS.root}/support/v8_runner.js",
          encoding: "UTF-8"
        )
      else
        current_runtime = ExecJS.runtime.name
        message = <<~MSG
          You have set `profile_server_rendering_js_code` to true, but the current execjs runtime is #{current_runtime}.
          ExecJS profiler only supports Node.js (V8) or V8 runtimes.
          You can set the runtime by setting the `EXECJS_RUNTIME` environment variable to either `Node` or `V8`.
        MSG
        raise ReactOnRailsPro::Error, message
      end
    end

    def node_renderer?
      ReactOnRailsPro.configuration.server_renderer == "NodeRenderer"
    end

    private

    def setup_assets_to_copy
      self.assets_to_copy = (Array(assets_to_copy) if assets_to_copy.present?)
    end

    def configure_default_url_if_not_provided
      self.renderer_url = renderer_url.presence || DEFAULT_RENDERER_URL
    end

    def validate_url
      URI(renderer_url)
    rescue URI::InvalidURIError => e
      message = "Unparseable ReactOnRailsPro.config.renderer_url #{renderer_url} provided.\n" \
                "#{e.message}"
      raise ReactOnRailsPro::Error, message
    end

    def validate_remote_bundle_cache_adapter
      if !remote_bundle_cache_adapter.nil? && !remote_bundle_cache_adapter.is_a?(Module)
        raise ReactOnRailsPro::Error, "config.remote_bundle_cache_adapter can only have a module or class assigned"
      end

      return unless remote_bundle_cache_adapter.is_a?(Module)

      unless remote_bundle_cache_adapter.methods.include?(:build)
        raise ReactOnRailsPro::Error,
              "config.remote_bundle_cache_adapter must have a class method named 'build'"
      end

      unless remote_bundle_cache_adapter.methods.include?(:fetch)
        raise ReactOnRailsPro::Error,
              "config.remote_bundle_cache_adapter must have a class method named 'fetch'" \
              "which takes a single named String parameter 'zipped_bundles_filename'" \
              "and returns the zipped file as a string if fetch attempt is successful & nil if not"
      end

      unless remote_bundle_cache_adapter.methods.include?(:upload) # rubocop:disable Style/GuardClause
        raise ReactOnRailsPro::Error,
              "config.remote_bundle_cache_adapter must have a class method named 'upload'" \
              "which takes a single named Pathname parameter 'zipped_bundles_filepath' & returns nil"
      end
    end

    def setup_renderer_password
      return if renderer_password.present?

      uri = URI(renderer_url)
      self.renderer_password = uri.password
    end
  end
end
